File: cpp/LeetCode/No1010.cpp
/*
You are given a list of songs where the ith song has a duration of time[i] seconds.

Return the number of pairs of songs for which their total duration in seconds is divisible by 60. 
Formally, we want the number of indices i, j such that i < j with (time[i] + time[j]) % 60 == 0.
*/
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int numPairsDivisibleBy60(vector<int>& time) {
        long result = 0;
        int map[60] = {0};
        for(int t : time){
            map[t%60]++;
        }
        for(int i=0;i<=30;i++){
            if(i==0 || i==30){
                if(map[i]>0){
                    result+=((long)map[i]*(long)(map[i]-1))/2;
                }
            }
            else{
                result+=(map[i]*map[60-i]);
            }
        }
        return result;
    }
};

int main(void){
    Solution solution;
    vector<int> time{30,20,150,100,40};
    cout << solution.numPairsDivisibleBy60(time) << endl;
    return 0;
}

File: python/LeetCode Quiz/No1010.py
'''
You are given a list of songs where the ith song has a duration of time[i] seconds.

Return the number of pairs of songs for which their total duration in seconds is divisible by 60. 
Formally, we want the number of indices i, j such that i < j with (time[i] + time[j]) % 60 == 0.
'''
from typing import List

class Solution:
    def numPairsDivisibleBy60(self, time: List[int]) -> int:
        table = dict()
        for t in time:
            if t%60 not in table:
                table[t%60] = 1
            else:
                table[t%60] += 1

        print(table)
        res = 0
        for t in table:
            if t==0 or t==30:
                res+=(table[t]*(table[t]-1))//2
            else:
                if 60-t in table:
                    res+=table[t]*table[60-t]
                    table[60-t] = 0
        return res
                
def main():
    solution = Solution()

    time = [30,20,150,100,40]
    print(solution.numPairsDivisibleBy60(time))

if __name__=='__main__':
    main()
